src/delivery_problems/mod.rs

```diff
@@ -0,0 +1,289 @@
+        let now = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("system time error");
+        let now = now.as_secs() * 1000;
+        problems
+            .iter()
+            .try_fold(HashMap::new(), |mut counts, problem| {
+                {
+                    let count = counts.entry(&problem.problem_type).or_insert(0);
+                    *count += 1;
+                    let limits = match problem.problem_type {
+                        ProblemType::HardBounce => &self.limits.hard,
+                        ProblemType::SoftBounce => &self.limits.soft,
+                        ProblemType::Complaint => &self.limits.complaint,
+                    };
+                    if is_bounce_violation(*count, problem.created_at, now, limits) {
+                        return match problem.problem_type {
+                            ProblemType::HardBounce => Err(AppErrorKind::BounceHardError {
+                                address: address.clone(),
+                                bounced_at: problem.created_at,
+                                bounce: problem.clone(),
+                            }.into()),
+                            ProblemType::SoftBounce => Err(AppErrorKind::BounceSoftError {
+                                address: address.clone(),
+                                bounced_at: problem.created_at,
+                                bounce: problem.clone(),
+                            }.into()),
+                            ProblemType::Complaint => Err(AppErrorKind::ComplaintError {
+                                address: address.clone(),
+                                bounced_at: problem.created_at,
+                                bounce: problem.clone(),
+                            }.into()),
+                        };
+                    }
+                }
+
+                Ok(counts)
+            }).map(|_| ())
+    }
+
+    /// Record a hard or soft bounce
+    /// against an email address.
+    pub fn record_bounce(
+        &self,
+        address: &EmailAddress,
+        bounce_type: BounceType,
+        bounce_subtype: BounceSubtype,
+    ) -> AppResult<()> {
+        self.auth_db
+            .create_bounce(address, From::from(bounce_type), From::from(bounce_subtype))?;
+        Ok(())
+    }
+
+    /// Record a complaint
+    /// against an email address.
+    pub fn record_complaint(
+        &self,
+        address: &EmailAddress,
+        complaint_type: Option<ComplaintFeedbackType>,
+    ) -> AppResult<()> {
+        let bounce_subtype = complaint_type.map_or(ProblemSubtype::Unmapped, |ct| From::from(ct));
+        self.auth_db
+            .create_bounce(address, ProblemType::Complaint, bounce_subtype)?;
+        Ok(())
+    }
+}
+
+unsafe impl<D> Sync for DeliveryProblems<D> where D: AuthDb {}
+
+fn is_bounce_violation(count: u8, created_at: u64, now: u64, limits: &[BounceLimit]) -> bool {
```

Comment:
This could be `is_limit_violation` or similar, to lean into the terminology change.